include/fiction/technology/sidb_cluster_hierarchy.hpp

```diff
@@ -0,0 +1,1080 @@
+     * @param loc_ext_pot The local external potential at the SiDB in the singleton cluster. Specifically, this is the
+     * sum of the local defect potential and the local external potential.
+     * @param base The simulation base.
+     * @param self_ptr Shared pointer to itself.
+     */
+    void initialize_singleton_cluster_charge_space(const double loc_pot_min, const double loc_pot_max,
+                                                   const double loc_ext_pot, const uint8_t base,
+                                                   const sidb_cluster_ptr& self_ptr) noexcept
+    {
+        assert(sidbs.size() == 1);
+
+        const uint64_t ix = *sidbs.cbegin();
+
+        for (const sidb_charge_state cs : sidb_charge_state_iterator<sidb_state_iter_dir::TO_NEGATIVE>{
+                 base == 3 ? sidb_charge_state::POSITIVE : sidb_charge_state::NEUTRAL})
+        {
+            charge_space.emplace(sidb_cluster_charge_state{self_ptr, cs, loc_ext_pot});
+        }
+
+        recv_ext_pot_bounds[ix][static_cast<uint8_t>(bound_direction::LOWER)] = loc_pot_min;
+        recv_ext_pot_bounds[ix][static_cast<uint8_t>(bound_direction::UPPER)] = loc_pot_max;
+    }
+    /**
+     * The bound on the potential an SiDB receives from outside this cluster is returned.
+     *
+     * @tparam bound The potential bound that is requested.
+     * @param i The SiDB index for which the potential bound is requested.
+     * @return The potential bound that is stored.
+     */
+    template <bound_direction bound>
+    constexpr inline double get_recv_ext_pot_bound(const sidb_ix i) noexcept
+    {
+        return recv_ext_pot_bounds.at(i).at(static_cast<uint8_t>(bound));
+    }
+    /**
+     * The bound on the potential an SiDB receives from outside this cluster is set.
+     *
+     * @tparam bound The potential bound that is set.
+     * @param i The SiDB index for which the potential bound is set.
+     * @param new_bound The new potential bound that is set.
+     */
+    template <bound_direction bound>
+    constexpr inline void set_recv_ext_pot_bound(const sidb_ix i, const double new_bound) noexcept
+    {
+        recv_ext_pot_bounds[i][static_cast<uint8_t>(bound)] = new_bound;
+    }
+    /**
+     * The bound on the potential an SiDB receives from outside this cluster is updated.
+     *
+     * @tparam bound The potential bound that is updated.
+     * @param i The SiDB index for which the potential bound is updated.
+     * @param diff The difference with which the potential bound is updated.
+     */
+    template <bound_direction bound>
+    constexpr inline void update_recv_ext_pot_bound(const sidb_ix i, const double diff) noexcept
+    {
+        recv_ext_pot_bounds[i][static_cast<uint8_t>(bound)] += diff;
+    }
+    /**
+     * Function to return the number of SiDBs contained in the cluster.
+     *
+     * @return The number of SiDBs contained in the cluster.
+     */
+    inline uint64_t size() const noexcept
```

Comment:
use a more expressive name.

Reply:
This is specifically made to be a shorthand so I don't need to write c->sidbs.size everywhere.

Reply:
Also, the number of children is never relevant in the algorithm so the cluster size can only mean one thing (not the number of children)

Reply:
I know, but ``size()`` can be anything, right? What about ``number_of_cluster_sidbs`` or similar?

Reply:
You are right. I think I'll go for `num_sidbs`, more in line with `num_cells`